homeassistant/components/airq/sensor.py

```diff
@@ -0,0 +1,324 @@
+        native_unit_of_measurement=ACTIVITY_BECQUEREL_PER_CUBIC_METER,
+        state_class=SensorStateClass.MEASUREMENT,
+        icon="mdi:radioactive",
+    ),
+    SensorEntityDescription(
+        key="temperature",
+        name="Temperature",
+        device_class=SensorDeviceClass.TEMPERATURE,
+        native_unit_of_measurement=TEMP_CELSIUS,
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    SensorEntityDescription(
+        key="tvoc",
+        name="VOC",
+        device_class=SensorDeviceClass.VOLATILE_ORGANIC_COMPOUNDS,
+        native_unit_of_measurement=CONCENTRATION_PARTS_PER_BILLION,
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    SensorEntityDescription(
+        key="tvoc_ionsc",
+        name="VOC (Industrial)",
+        device_class=SensorDeviceClass.VOLATILE_ORGANIC_COMPOUNDS,
+        native_unit_of_measurement=CONCENTRATION_PARTS_PER_BILLION,
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up sensor entities based on a config entry."""
+
+    coordinator = hass.data[DOMAIN][config.entry_id]
+    available_keys = list(coordinator.data.keys())
+
+    # Add sensors under warmup
+    status = coordinator.data["Status"]
+    if isinstance(status, dict):
+        warming_up_sensors = [
+            k for k, v in status.items() if "sensor still in warm up phase" in v
+        ]
+        available_keys.extend(warming_up_sensors)
+        _LOGGER.debug(
+            "Following %d sensors are warming up: %s",
+            len(warming_up_sensors),
+            ", ".join(warming_up_sensors),
+        )
+
+    # Filter out non-sensor keys and build a list of SensorEntityDescription objects
+    available_sensors = [
+        description for description in SENSOR_TYPES if description.key in available_keys
+    ]
+    _LOGGER.debug(
+        "Identified %d available sensors: %s",
+        len(available_sensors),
+        ", ".join([sensor.key for sensor in available_sensors]),
+    )
+
+    entities = [
+        AirQSensor(coordinator, description) for description in available_sensors
+    ]
+    async_add_entities(entities)
+
+
+class AirQSensor(CoordinatorEntity, SensorEntity):
+    """Representation of a Sensor."""
+
+    _attr_has_entity_name = True
+
+    def __init__(
+        self,
+        coordinator: AirQCoordinator,
+        description: SensorEntityDescription,
+    ) -> None:
+        """Initialize a single sensor."""
+        super().__init__(coordinator)
+        self.entity_description = description
+
+        self._attr_device_info = coordinator.device_info
+        self._attr_name = description.name
+        self._attr_unique_id = f"{coordinator.device_id}_{description.key}"
+        # the following two sensor units must be converted to %
+        self._factor = 0.1 if description.key in ["performance", "health"] else 1.0
```

Comment:
You can create your own dataclass description based on `SensorEntityDescription` which will contain a lambda function that returns the value of the sensor. In this function, you can transform your value, for example divide it. This is good practice here.
https://github.com/home-assistant/core/blob/ac3534cba120688b1d40d1b6bbc947322aa8c776/homeassistant/components/adguard/sensor.py#L25-L53

Reply:
Sounds good, thanks for the suggestion. 
I chose to expose a relatively lean lambda allowing to transform already retrieved value (while in `adguard` `value_fn` appears to also take on the job of retrieving the data from `AdGuardHome`). Hope it's ok.

Reply:
In such approach I am a little concerned with the need to duplicate the key of each sensor. Exposing the key in the lambda could be an option, which would land us here:

```python
@dataclass
class AirQEntityDescription(SensorEntityDescription):
    """Describes AirQ sensor entity. """

    value_fn: Callable[
        [AirQCoordinator, str], float | int | None
    ] = lambda coordinator, key: coordinator.data.get(key)

SENSOR_TYPES: list[AirQEntityDescription] = [
    AirQEntityDescription(
        key="nh3_MR100",
        name="Ammonia",
        native_unit_of_measurement=CONCENTRATION_MICROGRAMS_PER_CUBIC_METER,
        state_class=SensorStateClass.MEASUREMENT,
    ),
    AirQEntityDescription(
        key="health",
        name="Health Index",
        native_unit_of_measurement=PERCENTAGE,
        state_class=SensorStateClass.MEASUREMENT,
        icon="mdi:heart-pulse",
        value_fn=lambda coordinator, key: coordinator.data.get(key) / 10.0
    ),
]

@property
def native_value(self) -> float | int | None:
    """Return the value reported by the sensor."""
    return self.entity_description.value_fn(
        self.coordinator, self.entity_description.key
    )
```

However, I think I do not immediately see enough reasons to factor `coordinator.data.get(key)` out out of `native_value` into a lambda, since at present it is the step shared by all sensors. I would thus refrain from introducing `value_fn` as a function which retrieves desired value in the most general case given the coordinator, but instead view it as an optional transform which can be applied to selected (few) values. Perhaps, it can be renamed to `value_transform_fn` or smth.

Reply:
As an alternative:
```py
@property
def native_value(self) -> float | int | None:
    """Return the value reported by the sensor."""
    if self.entity_description.value_fn:
        return self.entity_description.value_fn(self.coordinator)
    return self.coordinator.get(self.entity_description.key)
```

But I think renaming to `value_transform_fn`  is also ok.

Reply:
Still, it's probably better to get data in a lambda function. 
Check out the implementation https://github.com/home-assistant/core/blob/dev/homeassistant/components/nextdns/sensor.py

You also need to check if your sensor data is updated in real time, as you may need set callback the coordinator as in the example above.

Reply:
Thanks for the example. 

> You also need to check if your sensor data is updated in real time

I am not 100% sure where and how to best check (i.e. assert in the code?). Yet I swapped the `native_value` property for a `callback` as in the example. I hope that's correct.

> Still, it's probably better to get data in a lambda function.

Latest commits make it so, but I am afraid that in this case I still find it hard to justify the duplication of the keys and repetition of the same call in every entity description instance (also in the example of nextdns). I probably miss something, in which case please let me know :)

Reply:
Look good now!